### PR TITLE
Track bone matrix updates for future morph usage

### PR DESCRIPTION
**Pull request by unknownbrackets**
See here for the explanation:
https://github.com/hrydgard/ppsspp/issues/7746#issuecomment-102743096

But basically morphing needs the bone matrices dirtied eventually.

-[Unknown]

### CODE REVIEW

File: GPU/GLES/GLES_GPU.cpp [hunk 1]
```diff
@@ -1915,15 +1928,18 @@ void GLES_GPU::Execute_Generic(u32 op, u32 diff) {
...
 		Flush();
-		const int num = gstate.boneMatrixNumber & 0x7F;
-		int mtxNum = num / 12;
-		uint32_t uniformsToDirty = DIRTY_BONEMATRIX0 << mtxNum;
-		if ((num - 12 * mtxNum) != 0) {
-			uniformsToDirty |= DIRTY_BONEMATRIX0 << ((mtxNum + 1) & 7);
-		}
 		shaderManager_->DirtyUniform(uniformsToDirty);
+	} else {
+		gstate_c.deferredVertTypeDirty |= uniformsToDirty;
```
Comment (hrydgard) [inline GPU/GLES/GLES_GPU.cpp]: Technically we could avoid computing uniformsToDirty at all in the case of softskinning and no morph, but not sure if it's worth the extra check.

**Comment by hrydgard** [comment]
Nice catch!

File: GPU/GLES/GLES_GPU.cpp (hunk 1 above)
Comment (unknownbrackets) [inline GPU/GLES/GLES_GPU.cpp]: No, we exactly cannot do that.

Example:

1. There is no morph.
2. The game updates a bone matrix.  Soft skinning does its magic.
3. Morph is turned on.  It was off before.
4. Draw happens.  Reuses previously set bone matrix.
5. Draw must have CORRECT bone matrix.

We have to compute the dirty flags always.  The only case we can avoid them is when we are sure that morph will **never** be used without first updating the bone matrices it references or turning off skinning.

-[Unknown]

Comment (hrydgard) [inline GPU/GLES/GLES_GPU.cpp]: Sorry, I see now! Thanks for the details.

**Comment by unknownbrackets** [comment]
Actually, I guess the other option is we could apply morph and skinning in vertex decode (e.g. in software.)  Would only need nrm and pos, right?

That might actually be faster since it would remove checks from the verttype and bone matrix gpu cmds...

-[Unknown]

**Comment by hrydgard** [comment]
Yes, that's certainly an option, although applying both at the same time will complicate the vertex decode quite a bit, I think... at least on ARM32 and x86-32 where we have some pretty bad lack of free registers already. If you want to spend some time on it I won't stop you though :)

**Comment by unknownbrackets** [comment]
Hmm.  On x86, it seems like we could essentially just do Jit_AnyS8Morph followed by Jit_WriteMatrixMul.  I think arm would be roughly the same.  Wouldn't need extra regs (long term), at least currently, right?

We currently don't cache the morph weights in regs at all afaik.

I'm not thinking to make skin/morph fast - I'm thinking simplest thing possible there.  I'm more concerned about speeding up the common case by removing checks for morph.

-[Unknown]

**Comment by hrydgard** [comment]
I was afraid that the morph code would overwrite the matrix constructed during weight loading, but if that's not an issue then that will work of course (didn't actually check if there was overlap).

**Comment by unknownbrackets** [comment]
Hmm, right, there's overlap with fpScratchReg4 and D4 / Q2.  But there are ways around that, I think.

I also wonder... maybe we could defer the DecodeVertsStep() until the bone matrix is updated?  That is, instead of flushing, decode?  That might cut down on matrix setup time in the vertexjit.

-[Unknown]

**Pull request merged**